### PR TITLE
Manual maintenance sprint 60

### PR DESCRIPTION
**Pull request by exatar**
# Manual Maintenance Sprint [NR]

Note: To create a PR using this template add the query parameter `template=manual_maintenance.md` to the merge request
creation URL (or simply copy this md file into the description)

# Maintenance tasks (to be completed by the assignee)

### Notes from Dec, 15 2023
Some tests are failing locally due to the cookie banner. A description is given in
https://jira.d-fine.dev/browse/DALA-3137

### Skipped updates

The following known issues need to be reviewed in case a compatible version is available. Add new known issues as they
appear.

- [x] JDK/JRE must remain at 17, upgrading it to 21 caused too many errors, also Kotlin 1.9.10 is not compatible with Java 21
- [x] kotlin to remain at 1.9.10 instead of 1.9.20, which caused error
- [x] eclipse-temurin exists in version 21 already but cannot be updated, as that breaks e2e tests in CI, we're using 17.
- [x] some sec fixes or updates to `package.json` in /frontend and /keycloak break the build:
  - frontend:
    - [x] axios > 1.6.2.. in the last MM axios was updated to 1.6.7 and it works. Nevertheless, one could still keep an eye on it
    - [x] keycloak-js > 22.0.5
    - [x] primevue > 3.44.0
    - [x] vue > 3.4.5
    - [x] @vue/tsconfig > 0.1.3 (Update "@vue/tsconfig" to >=0.2.0 introduces major changes in typescript rules (~500 TS Errors throughout the
      project and unresolved imports that are hard to fix))
    - [x]  cypress > 12.11.0 (Update Cypress to >= 12.12.0 introduces an issue with the usage of `$route` in component test
      `DatasetOverview.cy.ts`. Issue with Cypress has been created to hopefully resolve this from the side of Cypress:
      https://github.com/cypress-io/cypress/issues/26902)
  - keycloak:
    - [x] @zxcvbn-ts/language-common > 2.0.1 (issues in rebuilding keycloak Docker images)
    - [x] @zxcvbn-ts/language-en > 2.1.0(issues in rebuilding keycloak Docker images)
    - [x] @types/node > 20.10.8 causes issues with vite > 4.5.1 which causes build to fail
- [x] Update e2etests/Dockerfile update breaks the build
- [x] Update Ktlint to >= 49.0 breaks the ktlint tasks (issue described here: 
  https://github.com/JLLeitschuh/ktlint-gradle/issues/665 and possible fix here: 
  https://github.com/JLLeitschuh/ktlint-gradle/pull/667)
- [x] Update Postgres in Docker-compose.yml to 16.0 causes CD to fail. Postgres can't be upgraded to 16 as existing data is not compatible.
- [x] The docker-compose-plugin v.2.19.1 causes connection issues:
- [x] Check that it is still valid for `**/CompanyApi.kt', '**/CompanyDataController.kt` to be excluded from `config/detekt.yml`, 
      at latest once the refactoring of the APIs is done this must be reevaluated
### Gradle update

- [x] Execute `gradlew dependencyUpdates` to get a report on Dependencies with updates
- [x] Execute `refreshVersions` in Gradle tasks or `gradlew refreshVersions` to generate version suggestions in `versions.properties`
- [x] Update versions in `versions.properties`
  - [] as of 31.01.2024 plugin.org.springframework.boot=3.2.2 will break the CI, 3.2.1 works fine
  - [] here is the error message for future reference
    - tests/e2e/specs/infrastructure/VerifyDeployment.ts(10,38): error TS2339: Property 'health' does not exist on type 'ActuatorApi'. 
    - tests/e2e/specs/infrastructure/VerifyDeployment.ts(11,20): error TS18046: 'healthResponse' is of type 'unknown'.
- [x] Update the gradle wrapper: execute `gradle wrapper --gradle-version X.Y.Z`

### Dataland frontend

- [x] Update node version in `dataland-frontend/build.gradle.kts`
- [x] Update node packages: run the `updatepackages` script, e.g. by  `npm run updatepackages` to update versions in
  `package.json`
- [x] Run the `updatepackagelock`, e.g. by  `npm run updatepackagelock` script to update `package-lock.json` and check
  for security issues

### Dataland keycloak theme

- [x] Update node version in `dataland-keycloak/dataland_theme/login/build.gradle.kts`
- [x] Update node packages: run the `updatepackages` script, e.g. by  `npm run updatepackages` to update versions in
  `package.json`
- [x] Run the `updatepackagelock`, e.g. by  `npm run updatepackagelock` script to update `package-lock.json` and check
  for security issues

### Dataland automated QA service

- [x] Update package versions in `dataland-automated-qa-service/requirements.txt`

### Dockerfile updates

Update versions in the following dockerfiles

- [ ] `./dataland-api-key-manager/Dockerfile`
- [ ] `./dataland-api-key-manager/DockerfileBase`
- [ ] `./dataland-api-key-manager/DockerfileTest`
- [ ] `./dataland-automated-qa-service/Dockerfile`
- [ ] `./dataland-automated-qa-service/DockerfileBase`
- [ ] `./dataland-automated-qa-service/DockerfileTest`
- [ ] `./dataland-backend/Dockerfile`
- [ ] `./dataland-backend/DockerfileBase`
- [ ] `./dataland-backend/DockerfileTest`
- [ ] `./dataland-batch-manager/Dockerfile`
- [ ] `./dataland-batch-manager/DockerfileBase`
- [ ] `./dataland-community-manager/Dockerfile`
- [ ] `./dataland-community-manager/DockerfileBase`
- [ ] `./dataland-community-manager/DockerfileTest`
- [ ] `./dataland-document-manager/Dockerfile`
- [ ] `./dataland-document-manager/DockerfileBase`
- [ ] `./dataland-document-manager/DockerfileTest`
- [ ] `./dataland-e2etests/Dockerfile`
- [ ] `./dataland-e2etests/DockerfileBase`
- [ ] `./dataland-frontend/Dockerfile`
- [ ] `./dataland-frontend/DockerfileTest`
- [ ] `./dataland-inbound-admin-proxy/Dockerfile`
- [ ] `./dataland-inbound-proxy/Dockerfile`
- [ ] `./dataland-internal-storage/Dockerfile`
- [ ] `./dataland-internal-storage/DockerfileBase`
- [ ] `./dataland-internal-storage/DockerfileTest`
- [ ] `./dataland-keycloak/Dockerfile`  (also update realm json files with new version)
- [ ] `./dataland-pgadmin/Dockerfile`
- [ ] `./dataland-qa-service/Dockerfile`
- [ ] `./dataland-qa-service/DockerfileBase`
- [ ] `./dataland-qa-service/DockerfileTest`
- [ ] `./dataland-rabbitmq/Dockerfile`
- [ ] `./dataland-inbound-admin-proxy/Dockerfile`
- [ ] `./dataland-inbound-proxy/Dockerfile`
- [ ] `./base-dockerfiles/DockerfileGradle`
- [ ] Update the versions of the external images for api-key-manager-db, backend-db, keycloak-db, internal-storage-db,
  document-manager-db, qa-service-db, community-manager-db and frontend-dev in `./docker-compose.yml`
- [ ] Check if there are any services in the `docker-compose.yml` file that have not gotten an update yet (e.g. a new
  service that is not covered by the tasks above)

## Dataland Monitoring

- [x] Go to the monitoring repository and execute the tasks described in the manual maintenance template there

## Server updates

Execute `sudo apt-get update && sudo apt-get upgrade` on

- [ ] dev1.dataland.com
- [ ] dev2.dataland.com
- [ ] dev3.dataland.com
- [ ] test.dataland.com
- [ ] letsencrypt.dataland.com
- [ ] monitoring.dataland.com
- [ ] (OPT) dataland.com

### ssh-keys maintenance

check that all ssh-keys are set and erased from people that have left

- [ ] dev1.dataland.com
- [ ] dev2.dataland.com
- [ ] dev3.dataland.com
- [ ] test.dataland.com
- [ ] letsencrypt.dataland.com
- [ ] monitoring.dataland.com
- [ ] (OPT) dataland.com

## Check keycloak automatic logout if inactive

- [ ] Check that you are automatically logged out if you are idle for too long and also get notified about this by a
  pop-up. Timeouts are defined as `TIME_BEFORE_REFRESH_TOKEN_EXPIRY_TO_DISPLAY_SESSION_WARNING_IN_MS`
  in `./dataland-frontend/src/utils/Constants.ts` and `ssoSessionIdleTimeout`
  in `./dataland-keycloak/realms/datalandsecurity-realm.json`.

## Check e-mails sent by keycloak

- [ ] Check that the verification e-mail is sent and displayed correctly
- [ ] Check that the reset password e-mail is sent and displayed correctly
- [ ] Check that the password can be reset
- [ ] Check that account linking via e-mail verification works correctly
- [ ] Check that account linking via username & password verification works correctly

## Check e-mails sent by backend

- [ ] Send an invitation (data) request from one of the dev servers and check if the e-mail to dataland@d-fine.com
  contains the right attachments and is displayed correctly.

## Check RabbitMQ dead letter queue and disk space

- [ ] RabbitMQ does need at least 768MB of free disk space to operate. `ssh` into all servers and check the available
  disk space with `df` command. If the open disk space is close to the minimum requirement, clear up disk space
  with `sudo docker image prune --all`.
- [ ] On all environments, no new messages should have been added to the dead letter queue since the last manual
  maintenance. If new messages have appeared this does need to be investigated. The dead letter queue can be accessed
  and messages on it read in the RabbitMQ GUI. Access it by port-forwarding port `6789` from the server and then
  accessing the GUI at `localhost:6789/rabbitmq`. After login, the dead letter queue can be found at Queues &rarr;
  deadLetterQueue &rarr; Get message.

## Check that links to external webpages are working
- [ ] Go to the frameworks/upload page and check that the link to WWF Pathway to Paris is still working as intended.

## Check that the main branch has no sonar issues
- [ ] Go to the sonar report summary of the main branch and verify that there are no sonar findings. If there are sonar 
  findings fix them.

## Conclusion

- [ ] After updating all components check if everything still works
- [ ] The new version is deployed to a dev server using this branch
    - [ ] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
    - [ ] It's verified that everything seems to be working fine by manually using the website
    - [ ] All implemented Social Logins have been tested manually in the UI
- [ ] This template has been updated to reflect the latest state of tasks required and known issues with upgrades
- [ ] The Merge Request commit message needs to contain 'manual maintenance' to satisfy the CI maintenance check in
  future commits

# Review (to be completed by the reviewer)

- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github.
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described above
- [ ] Documentation is updated as required
- [ ] The automated deployment is updated if required
- [x] The new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] It's verified that everything seems to be working fine by manually using the website
  - [x] All implemented Social Logins have been tested manually in the UI
  - [x] The "mailto:*" buttons ("I am interested" and "Get in touch") on the landing page work.
  - [x] Go to the swagger-UI, authorize, run a "GET" request to the companies endpoint and assure that your
    authorization has worked by assuring that you get a 200 response
- [ ] If any work on the UI is to be merged, those changes were also documented in the Figma
- [x] The local Dev stack still works: execute `startDevelopmentStack.sh` and execute `npm run testpipeline`
  and `npm run testcomponent` in dataland-frontend
- [ ] After(!) the cypress tests have passed locally, execute the backend-e2e-tests `./gradlew dataland-e2etests:test`
- [x] Locally: Go to the swagger-UI, authorize, run a "GET" request to the companies endpoint and assure that your
  authorization has worked by assuring that you get a 200 response
- [ ] It is assured that deploying this feature branch over the current main does not break anything
  - [x] Deploy the version of main currently active on prod to a dev server with `Reset non-user related Docker Volumes & Re-populate` turned on
  - [x] Verify that the CD run is green
  - [x] Migrate the data from prod to the dev server using `./migrateData.sh dataland.com <SOURCE_API_KEY> <TARGET> <TARGET_API_KEY>`
  - [x] Deploy the feature branch to the same server with `Reset non-user related Docker Volumes & Re-populate` turned off
  - [x] Verify that the CD run is green
  - [x] Verify that everything seems to be working fine by manually using the website
- [ ] Merge using Squash Commit. The Merge Commit Message needs to contain "Manual Maintenance"
- [ ] After merge check SonarQube state of main branch at https://sonarcloud.io/summary/new_code?id=d-fine_Dataland. 
  The full scan might reveal new issues (e.g. deprecation) on old code which is generally not detected on the branch.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ appear.
 - [ ] eclipse-temurin exists in version 21 already but cannot be updated, as that breaks e2e tests in CI, we're using 17.
 - [ ] some sec fixes or updates to `package.json` in /frontend and /keycloak break the build:
   - frontend:
-    - [ ] axios > 1.6.2
+    - [ ] axios > 1.6.2.. in the last MM axios was updated to 1.6.7 and it works. Nevertheless, one could still keep an eye on it
```
Comment (PaulChristiansen) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:24]: Is 1.6.7 the latest axios version? If that is the case (and if everything is working as expected), I would put that in the focus here, for example by writing:

"(axios was updated to latest version 1.6.7 in sprint 60; however, it caused issues before when updating to > 1.6.2, so keep an eye on it)"

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 2]
```diff
@@ -76,28 +80,37 @@ Update versions in the following dockerfiles
...
+- [ ] `./dataland-automated-qa-service/DockerfileBase`
+- [ ] `./dataland-automated-qa-service/DockerfileTest`
 - [ ] `./dataland-backend/Dockerfile`
 - [ ] `./dataland-backend/DockerfileBase`
 - [ ] `./dataland-backend/DockerfileTest`
+- [ ] `./dataland-batch-manager/Dockerfile`
+- [ ] `./dataland-batch-manager/DockerfileBase`
+- [ ] `./dataland-community-manager/Dockerfile`
+- [ ] `./dataland-community-manager/DockerfileBase`
+- [ ] `./dataland-community-manager/DockerfileTest`
```
Comment (PaulChristiansen) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:93]: Why were these added Dockerfiles not listed before? Is this just for the sake of completeness?

File: dataland-frontend/src/components/forms/parts/elements/derived/ProcurementCategoryFormElement.vue [hunk 3]
```diff
@@ -141,10 +141,9 @@ export default defineComponent({
...
      */
     setPreSelectedCountries() {
-      return this.allCountries.filter(
-        (el) =>
-          // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,no-prototype-builtins
-          this.procurementCategories[this.name]?.numberOfSuppliersPerCountryCode?.hasOwnProperty(el.value),
+      return this.allCountries.filter((el) =>
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,no-prototype-builtins
+        this.procurementCategories[this.name]?.numberOfSuppliersPerCountryCode?.hasOwnProperty(el.value),
       );
```
Comment (PaulChristiansen) [inline dataland-frontend/src/components/forms/parts/elements/derived/ProcurementCategoryFormElement.vue:147]: Is this something caused by the linter?

Comment (exatar) [reply dataland-frontend/src/components/forms/parts/elements/derived/ProcurementCategoryFormElement.vue:147]: Yes

File: dataland-frontend/tests/component/components/pages/LandingPage.cy.ts [hunk 4]
```diff
@@ -111,9 +111,7 @@ function getLandingPageSection(sectionTitle: string): Section {
  * @returns the filename of the found image
  */
 function getSingleImageNameInSection(sectionTitle: string): string {
-  return assertDefined(getLandingPageSection(sectionTitle)?.image?.[0])
-    .split("/")
-    .slice(-1)[0];
+  return assertDefined(getLandingPageSection(sectionTitle)?.image?.[0]).split("/").slice(-1)[0];
```
[the remark this answers is not in the document]
Comment (exatar) [reply dataland-frontend/tests/component/components/pages/LandingPage.cy.ts:114]: Yes

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 2 above)
Comment (exatar) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:93]: Yes

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -42,6 +42,7 @@ kotlin = "1.9.10"
 okhttp3 = "4.12.0"
 
 moshi = "1.15.0"
+## ⬆ = "1.15.1"
```
Comment (PaulChristiansen) [inline gradle/libs.versions.toml:46]: What about this moshi update?

File: versions.properties [hunk 6]
```diff
@@ -86,7 +95,8 @@ version.kotlin=1.9.10
 ## unused
 version.junit.jupiter=5.10.0
 
-version.io.gitlab.arturbosch.detekt..detekt-cli=1.23.4
+version.io.gitlab.arturbosch.detekt..detekt-cli=1.23.5
+##                                  # available=1.23.5
```
Comment (PaulChristiansen) [inline versions.properties:99]: Here

File: versions.properties [hunk 7]
```diff
@@ -115,24 +125,28 @@ version.ch.qos.logback..logback-classic=1.4.14
...
 
 plugin.org.springdoc.openapi-gradle-plugin=1.8.0
 
 plugin.org.sonarqube=4.4.1.3373
 
 plugin.org.openapi.generator=7.2.0
 
-plugin.org.jlleitschuh.gradle.ktlint=12.0.3
+plugin.org.jlleitschuh.gradle.ktlint=12.1.0
+##                       # available=12.1.0
```
Comment (PaulChristiansen) [inline versions.properties:137]: Here

File: versions.properties [hunk 8]
```diff
@@ -115,24 +125,28 @@ version.ch.qos.logback..logback-classic=1.4.14
...
 
-plugin.io.gitlab.arturbosch.detekt=1.23.4
+plugin.io.gitlab.arturbosch.detekt=1.23.5
+##                     # available=1.23.5
```
Comment (PaulChristiansen) [inline versions.properties:141]: Here

File: versions.properties [hunk 9]
```diff
@@ -115,24 +125,28 @@ version.ch.qos.logback..logback-classic=1.4.14
...
 
 plugin.com.gorylenko.gradle-git-properties=2.4.1
 
 plugin.com.github.node-gradle.node=7.0.1
 
 plugin.com.github.jk1.dependency-license-report=2.5
 
-plugin.com.github.ben-manes.versions=0.50.0
+plugin.com.github.ben-manes.versions=0.51.0
+##                       # available=0.51.0
```
Comment (PaulChristiansen) [inline versions.properties:150]: Here

File: gradle/libs.versions.toml (hunk 5 above)
Comment (exatar) [reply gradle/libs.versions.toml:46]: Since it is unused accroding to the versions.properties file we skipped the update.

File: versions.properties [hunk 10]
```diff
@@ -30,32 +30,39 @@ version.org.freemarker..freemarker=2.3.32
...
 version.org.mockito.kotlin..mockito-kotlin=5.2.1
 
 ## unused
 version.org.json..json=20230618
 
 version.org.flywaydb..flyway-database-postgresql=10.4.1
+##                                   # available=10.5.0
+##                                   # available=10.6.0
+##                                   # available=10.7.0
+##                                   # available=10.7.1
```
[the remark this answers is not in the document]
Comment (PaulChristiansen) [reply versions.properties:60]: No, I think the comments above for flyway-database-postgresql can stay as they are - also at this place. However, where I've added the comment "Here" below I think the comments with the available versions can be removed as they only provide redundant information

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 1 above)
Comment (exatar) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:24]: Adjusted the comment accordingly and will part of the next push

File: versions.properties (hunk 10 above)
[the remark this answers is not in the document]
Comment (exatar) [reply versions.properties:60]: done

File: versions.properties [hunk 11]
```diff
@@ -76,7 +83,9 @@ version.okhttp3=4.12.0
 ## unused
 version.moshi=1.15.0
 
-version.mockito=5.8.0
+version.mockito=5.10.0
+##  # available=5.9.0
+##  # available=5.10.0
```
[the remark this answers is not in the document]
Comment (exatar) [reply versions.properties]: done

File: versions.properties (hunk 8 above)
Comment (exatar) [reply versions.properties:137]: done

File: gradle/libs.versions.toml (hunk 5 above)
Comment (exatar) [reply gradle/libs.versions.toml:46]: pushed a new version where moshi is updated to 1.15.1
